### PR TITLE
Fix: Posts widget - query is not being imported correctly when importing a Kit [ED-4687]

### PR DESCRIPTION
**Pull request by mark-elementor**

### CODE REVIEW

File: core/base/document.php [hunk 1]
```diff
@@ -1172,6 +1174,33 @@ public static function on_import_replace_dynamic_content( $config, $map_old_new_
...
+	 * @since 3.8.0
+	 * @access public
+	 *
+	 * @param array $config
+	 * @param array $new_ids_map
+	 * @param array|null $controls
+	 *
+	 * @return array Element data.
+	 */
+	public static function updated_on_import_replace_dynamic_content( array $config, array $new_ids_map, $controls = null ) : array {
```
Comment (ManorHazaz) [inline core/base/document.php:1189]: Maybe use $data or any other general naming?
So in the future, we can use this function for more diverse uses.

```suggestion
	public static function updated_on_import_replace_dynamic_content( array $config, array $data, $controls = null ) : array {
```

File: core/base/document.php [hunk 2]
```diff
@@ -1151,14 +1151,16 @@ public function delete() {
...
 	 * @param array $config
-	 *
 	 * @param array $map_old_new_post_ids
+	 *
+	 * @deprecated 3.8.0 Use `::updated_on_import_replace_dynamic_content()` instead.
 	 */
 	public static function on_import_replace_dynamic_content( $config, $map_old_new_post_ids ) {
+		Plugin::$instance->modules_manager->get_modules( 'dev-tools' )->deprecation->deprecated_function( __METHOD__, '3.8.0', __CLASS__ . '::updated_on_import_replace_dynamic_content()' );
+
 		foreach ( $config as &$element_config ) {
```
Comment (ManorHazaz) [inline core/base/document.php:1164]: Refer this suggestion to all the lines from 1164-1174
Maybe use the newly created function instead of duplicating the code?
```suggestion
		$data = [ 'post_ids' => $map_old_new_post_ids, ];
		return static::updated_on_import_replace_dynamic_content( $config, $data ); 
```

File: includes/base/controls-stack.php [hunk 3]
```diff
@@ -1984,11 +1984,30 @@ public function print_template() {
...
+		Plugin::$instance->modules_manager->get_modules( 'dev-tools' )->deprecation->deprecated_function( __METHOD__, '3.8.0', __CLASS__ . '::updated_on_import_replace_dynamic_content()' );
+
+		return $config;
+	}
+
+	/**
+	 * On import replace dynamic content.
+	 *
+	 * @since 3.8.0
+	 * @access public
```
Comment (ManorHazaz) [inline includes/base/controls-stack.php:2002]: Redundant.
```suggestion
```

File: includes/base/element-base.php [hunk 4]
```diff
@@ -619,6 +619,47 @@ public static function on_import_replace_dynamic_content( $config, $map_old_new_
...
+	 * @param array $config
+	 * @param array $new_ids_map
+	 * @param array|null $controls
+	 *
+	 * @return array Element data.
+	 */
+	public static function updated_on_import_replace_dynamic_content( array $config, array $new_ids_map, $controls = null ) : array {
+		$tags_manager = Plugin::$instance->dynamic_tags;
+
+		if ( empty( $config['settings'] ) || empty( $config['settings'][ $tags_manager::DYNAMIC_SETTING_KEY ] ) ) {
```
Comment (ManorHazaz) [inline includes/base/element-base.php:637]: Are you sure about the first condition? Looks like a mistake...

File: includes/base/element-base.php [hunk 5]
```diff
@@ -619,6 +619,47 @@ public static function on_import_replace_dynamic_content( $config, $map_old_new_
...
+			$tag_instance = $tags_manager->create_tag( $tag_config['id'], $tag_config['name'], $tag_config['settings'] );
+
+			if ( ! $tag_instance ) {
+				continue;
+			}
+
+			if ( ! method_exists( $tag_instance, 'updated_on_import_replace_dynamic_content' ) ) {
+				// on_import_replace_dynamic_content() is deprecated since 3.8.0
+				// TODO: Remove this check in the future.
+				$tag_config = $tag_instance->on_import_replace_dynamic_content( $tag_config, $new_ids_map );
```
Comment (ManorHazaz) [inline includes/base/element-base.php:652]: I think you should just pass 'map_old_new_post_ids' so it will work correctly.
Am I wrong?

File: includes/base/element-base.php [hunk 6]
```diff
@@ -619,6 +619,47 @@ public static function on_import_replace_dynamic_content( $config, $map_old_new_
...
+	 * @since 3.8.0
+	 * @access public
+	 *
+	 * @param array $config
+	 * @param array $new_ids_map
+	 * @param array|null $controls
+	 *
+	 * @return array Element data.
+	 */
+	public static function updated_on_import_replace_dynamic_content( array $config, array $new_ids_map, $controls = null ) : array {
```
Comment (ManorHazaz) [inline includes/base/element-base.php:634]: Same as above?
```suggestion
	public static function updated_on_import_replace_dynamic_content( array $config, array $data, $controls = null ) : array {
```

File: tests/phpunit/elementor/core/base/test-document.php [hunk 7]
```diff
@@ -90,7 +90,7 @@ public function test_on_import_replace_dynamic_content() {
 			'post_id' => $post->ID,
 		] );
 
-		$map_old_new_post_ids = [ '35' => '36', '37' => '38' ];
+		$new_ids_map = [ '35' => '36', '37' => '38' ];
```
Comment (ManorHazaz) [inline tests/phpunit/elementor/core/base/test-document.php:93]: $new_ids_map was associative array [ post_ids => ..., term_ids => ..., ] at the beginning, why did you change it and still used the same name?
I probably missed the part where you filter the array.

**Review by ghost — CHANGES_REQUESTED** [review]

File: app/modules/import-export/processes/import.php [hunk 8]
```diff
@@ -461,16 +460,26 @@ private function get_default_settings_include() {
 		return [ 'templates', 'plugins', 'content', 'settings' ];
 	}
 
+	/**
+	 * Get the data that requires updating/replacement when imported.
+	 *
+	 * @return array
+	 */
+	protected function get_imported_data_replacements() : array {
```
[the remark this answers is not in the document]
Comment (mark-elementor) [reply app/modules/import-export/processes/import.php:468]: Done

File: includes/base/controls-stack.php (hunk 3 above)
Comment (mark-elementor) [reply includes/base/controls-stack.php:2002]: Done

File: includes/base/element-base.php (hunk 4 above)
Comment (mark-elementor) [reply includes/base/element-base.php:637]: Changed.

File: core/base/document.php (hunk 1 above)
Comment (mark-elementor) [reply core/base/document.php:1189]: Changed

File: includes/base/element-base.php (hunk 5 above)
Comment (mark-elementor) [reply includes/base/element-base.php:652]: You are right. Changed.

File: includes/base/element-base.php (hunk 6 above)
Comment (mark-elementor) [reply includes/base/element-base.php:634]: Changed.

File: tests/phpunit/elementor/core/base/test-document.php (hunk 7 above)
Comment (mark-elementor) [reply tests/phpunit/elementor/core/base/test-document.php:93]: Changed back.

File: app/modules/import-export/processes/import.php [hunk 9]
```diff
@@ -461,16 +460,26 @@ private function get_default_settings_include() {
...
 	 * Handle the replacement of all the dynamic content of the elements that probably have been changed during the import.
-	 *
-	 * @param array $map_old_new_post_ids Mapped array of old and new post ids.
 	 */
-	private function save_elements_of_imported_posts( array $map_old_new_post_ids ) {
+	private function save_elements_of_imported_posts() {
 		foreach ( $this->documents_elements as $new_id => $document_elements ) {
 			$document = Plugin::$instance->documents->get( $new_id );
-			$updated_elements = $document->on_import_replace_dynamic_content( $document_elements, $map_old_new_post_ids );
+			$updated_elements = $document->updated_on_import_replace_dynamic_content( $document_elements, $this->get_imported_data_replacements() );
```
[the remark this answers is not in the document]
Comment (mark-elementor) [reply app/modules/import-export/processes/import.php]: Changed.

File: app/modules/import-export/processes/import.php [hunk 10]
```diff
@@ -461,16 +460,26 @@ private function get_default_settings_include() {
 		return [ 'templates', 'plugins', 'content', 'settings' ];
 	}
 
+	/**
+	 * Get the data that requires updating/replacement when imported.
+	 *
+	 * @return array
```
Comment (ManorHazaz) [inline app/modules/import-export/processes/import.php:466]: Just for your knowledge, you don't have to do it, it might be helpful with IDE autocomplete.
```suggestion
	 * @return array{post_ids: array, term_ids: array}
```

File: app/modules/import-export/processes/import.php [hunk 11]
```diff
@@ -461,16 +460,26 @@ private function get_default_settings_include() {
...
 	 * Handle the replacement of all the dynamic content of the elements that probably have been changed during the import.
-	 *
-	 * @param array $map_old_new_post_ids Mapped array of old and new post ids.
 	 */
-	private function save_elements_of_imported_posts( array $map_old_new_post_ids ) {
+	private function save_elements_of_imported_posts() {
 		foreach ( $this->documents_elements as $new_id => $document_elements ) {
 			$document = Plugin::$instance->documents->get( $new_id );
-			$updated_elements = $document->on_import_replace_dynamic_content( $document_elements, $map_old_new_post_ids );
+			$updated_elements = $document->on_import_update_ids( $document_elements, $this->get_imported_data_replacements() );
```
Comment (ManorHazaz) [inline app/modules/import-export/processes/import.php:482]: Maybe?
```suggestion
			$updated_elements = $document->on_import_update_dynamic_content( $document_elements, $this->get_imported_data_replacements() );
```

File: core/base/document.php [hunk 12]
```diff
@@ -1151,14 +1151,16 @@ public function delete() {
...
 	 * @since 3.6.0
 	 *
 	 * @param array $config
-	 *
 	 * @param array $map_old_new_post_ids
+	 *
+	 * @deprecated 3.8.0 Use `::on_import_update_ids()` instead.
 	 */
 	public static function on_import_replace_dynamic_content( $config, $map_old_new_post_ids ) {
+		Plugin::$instance->modules_manager->get_modules( 'dev-tools' )->deprecation->deprecated_function( __METHOD__, '3.8.0', __CLASS__ . '::on_import_update_ids()' );
```
Comment (ManorHazaz) [inline core/base/document.php:1162]: Same, and in any other place.
```suggestion
		Plugin::$instance->modules_manager->get_modules( 'dev-tools' )->deprecation->deprecated_function( __METHOD__, '3.8.0', __CLASS__ . '::on_import_update_dynamic_content()' );
```

File: core/base/document.php [hunk 13]
```diff
@@ -1172,6 +1174,33 @@ public static function on_import_replace_dynamic_content( $config, $map_old_new_
 		return $config;
 	}
 
+	/**
+	 * On import update post and term IDs.
+	 *
+	 * @since 3.8.0
+	 * @access public
```
Comment (ManorHazaz) [inline core/base/document.php:1181]: Remove.

File: includes/base/controls-stack.php [hunk 14]
```diff
@@ -1984,11 +1984,29 @@ public function print_template() {
...
+	}
+
+	/**
+	 * On import update post and term IDs.
+	 *
+	 * @since 3.8.0
+	 *
+	 * @param array $config
+	 * @param array $data
+	 * @param array|null $controls
```
Comment (ManorHazaz) [inline includes/base/controls-stack.php:2005]: Maybe explain the parameters and their uses?

**Review by ghost — DISMISSED** [review]

File: core/base/document.php (hunk 12 above)
Comment (mark-elementor) [reply core/base/document.php:1162]: Done.

File: core/base/document.php (hunk 13 above)
Comment (mark-elementor) [reply core/base/document.php:1181]: Removed.

File: app/modules/import-export/processes/import.php (hunk 11 above)
Comment (mark-elementor) [reply app/modules/import-export/processes/import.php:482]: Done.

File: app/modules/import-export/processes/import.php (hunk 10 above)
Comment (mark-elementor) [reply app/modules/import-export/processes/import.php:466]: True.
In this case won't really help in anything, but good to start using it as a habit.

File: includes/base/controls-stack.php (hunk 14 above)
Comment (mark-elementor) [reply includes/base/controls-stack.php:2005]: Done.

**Review by ghost — APPROVED** [review]
